nixos/modules/services/web-servers/keter/default.nix

```diff
@@ -0,0 +1,163 @@
+                      port 81 is set to default to allow nginx to reverse proxy into keter,
+                      so that nginx can do ssl management.
+                      You want that ip-from-header in the nginx setup case
+                      so it's not set to 127.0.0.1.
+                      Keep in mind root is overwritten by keterRoot.
+                      which is required for the nix derivation.
+                      '';
+    };
+
+    bundle = {
+      appName = lib.mkOption {
+        type = lib.types.str;
+        default = "myapp";
+        description = ''The name keter assigns to this bundle'';
+      };
+      executable = lib.mkOption {
+        type = lib.types.path;
+        description = ''The executable to be run'';
+      };
+
+      domain = lib.mkOption {
+        type = lib.types.str;
+        default = "example.com";
+        description = ''The domain keter will bind to'';
+      };
+
+      # these won't work with standard nix options because they have
+      # to be put in a tarball, so we use a runtime script to load settings.
+      publicScript = lib.mkOption {
+        type = lib.types.str;
+        default = "";
+        description = ''
+          Allows loading of public environment variables,
+          these are emitted to the log so it shouldn't contain secrets.
+          For example:
+          ```
+          ADMIN_EMAIL=hi@example.com
+          ```
+        '';
+      };
+
+      secretScript = lib.mkOption {
+        type = lib.types.str;
+        default = "";
+        description = ''
+          Allows loading of private environment variables,
+          for example:
+          ```
+          MY_AWS_KEY=$(cat /run/keys/AWS_ACCESS_KEY_ID)
+          ```
+        '';
+      };
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable (
+    let
+      incoming = "${cfg.keterRoot}/incoming";
+
+
+      globalKeterConfigFile = pkgs.writeTextFile {
+        name = "keter-config.yml";
+        text = (lib.generators.toYAML { } (cfg.globalKeterConfig // { root = cfg.keterRoot; }));
+      };
+
+      # If things are expected to change often, put it in the bundle!
+      bundle = pkgs.callPackage ./bundle.nix
+        (cfg.bundle // { inherit executable; });
+
+      # This indirection is required to ensure the nix path
+      # gets copied over to the target machine in remote deployments.
+      # Furthermore, it's important that we use exec to
+      # run the binary otherwise we get process leakage.
+      executable = pkgs.writeShellScript "bundle-wrapper" ''
+        set -e
+        ${cfg.bundle.secretScript}
+        set -xe
+        ${cfg.bundle.publicScript}
+        exec ${cfg.bundle.executable}
+      '';
+
+    in
+    {
+      # This is keter, which loads our app on deploy, and checks if it is healthy
+      # be very careful putting anything in here that changes,
+      # any change to the systemd unit can cause brief windows of downtime
+      # because nixos has to restart the unit.
```

Comment:
Like what? If the keter package gets updated that seems natural to me that it'd restart it.

Reply:
Yes.. you'd say so...
this was an internal comment disgurgaging against merging an upgraded nixpkgs and large migrations in a single deploy cycle.
I deleted it.